File: typescript_test_crate/tests/11_state_var_reentry.ts
class StateVarReentry {
    private _state: string;
    private _state_stack: Array<any>;
    private _state_context: Record<string, any>;
    private _return_value: any;

    constructor() {
        this._state_stack = [];
        this._state_context = {  };
        this._return_value = null;
        this._state = "Counter";
        this._enter();
    }

    private _transition(target_state: string, exit_args: any = null, enter_args: any = null) {
        if (exit_args) {
            this._exit(...exit_args);
        } else {
            this._exit();
        }
        this._state = target_state;
        if (enter_args) {
            this._enter(...enter_args);
        } else {
            this._enter();
        }
    }

    private _change_state(target_state: string) {
        this._state = target_state;
    }

    private _dispatch_event(event: string, ...args: any[]) {
        const handler_name = `_s_${this._state}_${event}`;
        const handler = (this as any)[handler_name];
        if (handler) {
            return handler.apply(this, args);
        }
    }

    private _enter(...args: any[]) {
        if (this._state === "Counter") {
            this._state_context["count"] = 0;
        }
    }

    private _exit(...args: any[]) {
        // No exit handlers
    }

    public increment(): number {
        this._return_value = null
        this._dispatch_event("increment")
        return this._return_value
    }

    public get_count(): number {
        this._return_value = null
        this._dispatch_event("get_count")
        return this._return_value
    }

    public go_other() {
        this._dispatch_event("go_other");
    }

    public come_back() {
        this._dispatch_event("come_back");
    }

    private _s_Other_increment() {
        this._return_value = -1;
        return this._return_value;;
    }

    private _s_Other_get_count() {
        this._return_value = -1;
        return this._return_value;;
    }

    private _s_Other_come_back() {
        this._transition("Counter", null, null);
    }

    private _s_Counter_go_other() {
        this._transition("Other", null, null);
    }

    private _s_Counter_increment() {
        this._state_context["count"] = this._state_context["count"] + 1;
        this._return_value = this._state_context["count"];
        return this._return_value;;
    }

    private _s_Counter_get_count() {
        this._return_value = this._state_context["count"];
        return this._return_value;;
    }
}


function main() {
    console.log("=== Test 11: State Variable Reentry ===");
    const s = new StateVarReentry();

    // Increment a few times
    s.increment();
    s.increment();
    let count = s.get_count();
    if (count !== 2) {
        throw new Error(`Expected 2 after two increments, got ${count}`);
    }
    console.log(`Count before leaving: ${count}`);

    // Leave the state
    s.go_other();
    console.log("Transitioned to Other state");

    // Come back - state var should be reinitialized to 0
    s.come_back();
    count = s.get_count();
    if (count !== 0) {
        throw new Error(`Expected 0 after re-entering Counter (state var reinit), got ${count}`);
    }
    console.log(`Count after re-entering Counter: ${count}`);

    // Increment again to verify it works
    const result = s.increment();
    if (result !== 1) {
        throw new Error(`Expected 1 after increment, got ${result}`);
    }
    console.log(`After increment: ${result}`);

    console.log("PASS: State variables reinitialize on state reentry");
}

main();

File: python_test_crate/tests/11_state_var_reentry.py
from typing import Any, Optional, List, Dict, Callable

class StateVarReentry:
    def __init__(self):
        self._state_stack = []
        self._state_context = {}
        self._return_value = None
        self._state = "Counter"
        self._enter()

    def _transition(self, target_state, exit_args = None, enter_args = None):
        if exit_args:
            self._exit(*exit_args)
        else:
            self._exit()
        self._state = target_state
        if enter_args:
            self._enter(*enter_args)
        else:
            self._enter()

    def _change_state(self, target_state):
        self._state = target_state

    def _dispatch_event(self, event, *args):
        handler_name = f"_s_{self._state}_{event}"
        handler = getattr(self, handler_name, None)
        if handler:
            return handler(*args)

    def _enter(self, *args):
        if self._state == "Counter":
            self._state_context["count"] = 0


    def _exit(self, *args):
        # No exit handlers
        pass

    def increment(self) -> int:
        self._return_value = None
        self._dispatch_event("increment")
        return self._return_value

    def get_count(self) -> int:
        self._return_value = None
        self._dispatch_event("get_count")
        return self._return_value

    def go_other(self):
        self._dispatch_event("go_other")

    def come_back(self):
        self._dispatch_event("come_back")

    def _s_Counter_go_other(self):
        self._transition("Other", None, None)

    def _s_Counter_increment(self) -> int:
        self._state_context["count"] = self._state_context["count"] + 1
        self._return_value = self._state_context["count"]
        return

    def _s_Counter_get_count(self) -> int:
        self._return_value = self._state_context["count"]
        return

    def _s_Other_come_back(self):
        self._transition("Counter", None, None)

    def _s_Other_increment(self) -> int:
        self._return_value = -1
        return

    def _s_Other_get_count(self) -> int:
        self._return_value = -1
        return


def main():
    print("=== Test 11: State Variable Reentry ===")
    s = StateVarReentry()

    # Increment a few times
    s.increment()
    s.increment()
    count = s.get_count()
    assert count == 2, f"Expected 2 after two increments, got {count}"
    print(f"Count before leaving: {count}")

    # Leave the state
    s.go_other()
    print("Transitioned to Other state")

    # Come back - state var should be reinitialized to 0
    s.come_back()
    count = s.get_count()
    assert count == 0, f"Expected 0 after re-entering Counter (state var reinit), got {count}"
    print(f"Count after re-entering Counter: {count}")

    # Increment again to verify it works
    result = s.increment()
    assert result == 1, f"Expected 1 after increment, got {result}"
    print(f"After increment: {result}")

    print("PASS: State variables reinitialize on state reentry")

if __name__ == '__main__':
    main()

